src/clib/cool1d_multi_g-cpp.C

```diff
@@ -0,0 +1,1740 @@
+  std::vector<double> kpAl2O3(my_fields->grid_dimension[0]);
+  std::vector<double> kpreforg(my_fields->grid_dimension[0]);
+  std::vector<double> kpvolorg(my_fields->grid_dimension[0]);
+  std::vector<double> kpH2Oice(my_fields->grid_dimension[0]);
+  std::vector<double> kptot(my_fields->grid_dimension[0]);
+
+  // grain temperature
+
+  std::vector<double> gasSiM(my_fields->grid_dimension[0]);
+  std::vector<double> gasFeM(my_fields->grid_dimension[0]);
+  std::vector<double> gasMg2SiO4(my_fields->grid_dimension[0]);
+  std::vector<double> gasMgSiO3(my_fields->grid_dimension[0]);
+  std::vector<double> gasFe3O4(my_fields->grid_dimension[0]);
+  std::vector<double> gasAC(my_fields->grid_dimension[0]);
+  std::vector<double> gasSiO2D(my_fields->grid_dimension[0]);
+  std::vector<double> gasMgO(my_fields->grid_dimension[0]);
+  std::vector<double> gasFeS(my_fields->grid_dimension[0]);
+  std::vector<double> gasAl2O3(my_fields->grid_dimension[0]);
+  std::vector<double> gasreforg(my_fields->grid_dimension[0]);
+  std::vector<double> gasvolorg(my_fields->grid_dimension[0]);
+  std::vector<double> gasH2Oice(my_fields->grid_dimension[0]);
+
+  // Iteration mask
+
+  gr_mask_type anydust, interp;
+  std::vector<gr_mask_type> itmask_tab(my_fields->grid_dimension[0]);
+  // !#define CALCULATE_TGAS_SELF_CONSISTENTLY
+  // #ifdef CALCULATE_TGAS_SELF_CONSISTENTLY
+  int iter_tgas;
+  double tgas_err, tgas0;
+  // #endif /* NOT important */
+  //      debug
+  double edotunit;
+  int i_max;
+  gr_float d_max;
+
+  // \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\/////////////////////////////////
+  // =======================================================================
+
+  // Set flag for dust-related options
+
+  if ((my_chemistry->h2_on_dust > 0)  ||  (my_chemistry->dust_chemistry > 0)  || 
+      (my_chemistry->dust_recombination_cooling > 0))  {
+    anydust = MASK_TRUE;
+  } else {
+    anydust = MASK_FALSE;
+  }
+
+  // Set flag for needing interpolation variables
+
+  if ((my_chemistry->primordial_chemistry > 0)  ||  (my_chemistry->dust_chemistry > 0))  {
+    interp = MASK_TRUE;
+  } else {
+    interp = MASK_FALSE;
+  }
+  // Set log values of start and end of lookup tables
+
+  logtem0 = std::log(my_chemistry->TemperatureStart);
+  logtem9 = std::log(my_chemistry->TemperatureEnd);
+  dlogtem= (std::log(my_chemistry->TemperatureEnd) - std::log(my_chemistry->TemperatureStart))/(double)(my_chemistry->NumberOfTemperatureBins-1 );
+
+  // Set units
+
+  dom      = internalu.urho*(std::pow(internalu.a_value,3))/mh_local_var;
+  dom_inv  = 1./dom;
+  tbase1   = internalu.tbase1;
+  xbase1   = internalu.uxyz/(internalu.a_value*internalu.a_units);    // uxyz is [x]*a      = [x]*[a]*a'        '
+  dbase1   = internalu.urho*std::pow((internalu.a_value*internalu.a_units),3); // urho is [dens]/a^3 = [dens]/([a]*a')^3 '
+  coolunit = (std::pow(internalu.a_units,5) * std::pow(xbase1,2) * std::pow(mh_local_var,2)) / (std::pow(tbase1,3) * dbase1);
```

Comment:
This particular logic had been duplicated between 4 or 5 different Fortran files. I designed the `internalu` related functionality to factor out this exact functionality (I took care to preserve the order of operations to prevent drift in answer-test results). I think it is probably worth doing this before the logic is able to start diverging.

The following suggestion shows how to integrate the `internalu` logic (if you want to do this). If you choose to make this change, you should remove the declaration of `xbase1` and `dbase1` variables, and explicitly confirm that this doesn't break any tests. 

```suggestion
  dom      = internalu_calc_dom_(internalu);
  dom_inv  = 1./dom;
  tbase1   = internalu.tbase1;
  coolunit = internalu.coolunit;
```